infrastructure/infrastructure/src/main/java/alabno/wserver/WebSocketHandler.java

```diff
@@ -109,13 +125,211 @@ private void handleRetrieveResult(JsonParser parser, WebSocket conn) {
         // Read the JSON file
         String fileContent = studentJob.readPostProcessorOutput();
 
-        // Prepare output message
-        JSONObject msgobj = new JSONObject();
-        msgobj.put("type", "postpro_result");
-        msgobj.put("title", title);
-        msgobj.put("student", student);
-        msgobj.put("data", fileContent);
-        conn.send(msgobj.toJSONString());
+        // Holds annotation information for each student-submitted file
+        Map<String, List<AnnotationWrapper>> submissionFeedbackMap = generateSubmissionFeedbackMap(fileContent);
+
+        // Create a set of file names from the annotations in the JSON output
+        Set<String> uniqueFiles = new HashSet<>();
+        JsonParser jsonParser = new JsonParser(fileContent);
+        JsonArrayParser annotations = jsonParser.getArrayParser("annotations");
+        for (JsonParser a : annotations) {
+            String filePath = a.getString("filename");
+            uniqueFiles.add(filePath);
+        }
+
+        switch (subtype) {
+          case "postprocessor":
+            sendPostprocessorMsg(title, student, fileContent, conn);
+            break;
+          case "annotated":
```

Comment:
Like mentioned above, this may change later in favour of switching to 2 separate messages